workflow/controller/steps.go

```diff
@@ -260,7 +260,7 @@ func (woc *wfOperationCtx) executeStepGroup(ctx context.Context, stepGroup []wfv
 		if !childNode.Fulfilled() {
 			completed = false
 		} else if childNode.Completed() {
-			hasOnExitNode, onExitNode, err := woc.runOnExitNode(ctx, step.OnExit, step.Name, childNode.Name, stepsCtx.boundaryID, stepsCtx.tmplCtx)
+			hasOnExitNode, onExitNode, err := woc.runOnExitNode(ctx, step.OnExit, childNode.Name, childNodeID, stepsCtx.tmplCtx)
```

Comment:
On exit nodes should run under the boundary of the node that called them (the "exited" node)

Reply:
Before:
```yaml
STEP                  TEMPLATE  PODNAME                       DURATION  MESSAGE
 ✔ dag-diamond-wzqnc  diamond
 ├─✔ A                echo      dag-diamond-wzqnc-2374547985  6s
 ├─✔ A.onExit         oe        dag-diamond-wzqnc-1294891656  6s
 ├─✔ B                echo      dag-diamond-wzqnc-2324215128  6s
 ├─✔ C                echo      dag-diamond-wzqnc-2340992747  6s
 ├─✔ B.onExit         oe        dag-diamond-wzqnc-250094443   6s
 ├─✔ C.onExit         oe        dag-diamond-wzqnc-728915506   6s
 ├─✔ D                echo      dag-diamond-wzqnc-2424880842  6s
 └─✔ D.onExit         oe        dag-diamond-wzqnc-2186811989  6s
```

After:
```yaml
STEP                  TEMPLATE  PODNAME                       DURATION  MESSAGE
 ✔ dag-diamond-ztvr2  diamond
 ├─✔ A                echo      dag-diamond-ztvr2-3230871262  2s
 │ └─✔ onExit         oe        dag-diamond-ztvr2-1416350393  6s
 ├─✔ B                echo      dag-diamond-ztvr2-3214093643  6s
 │ └─✔ onExit         oe        dag-diamond-ztvr2-2959263954  6s
 ├─✔ C                echo      dag-diamond-ztvr2-3197316024  6s
 │ └─✔ onExit         oe        dag-diamond-ztvr2-3577844235  6s
 └─✔ D                echo      dag-diamond-ztvr2-3314759357  6s
   └─✔ onExit         oe        dag-diamond-ztvr2-2103494724  6s
```

Reply:
Note this is not only for cosmetic reasons: boundary parallelism is the main reason for making this change